hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/AbstractBatchEntitySelectFetchInitializer.java

```diff
@@ -89,12 +91,32 @@ protected void resolveKey(
 		}
 		else {
 			entityKey = new EntityKey( entityIdentifier, concreteDescriptor );
-
 			state = State.KEY_RESOLVED;
+			registerResolutionListener();
+		}
+	}
 
+	protected Object getExistingInstance(RowProcessingState rowProcessingState) {
+		if ( entityKey != null ) {
```

Comment:
I think this can be turned into an assertion as well instead of doing the if check.